include/kernels/dslash_clover_helper.cuh

```diff
@@ -184,63 +186,68 @@ namespace quda {
     constexpr NdegTwistCloverApply(const Arg &arg) : arg(arg) {}
     static constexpr const char* filename() { return KERNEL_FILE; }
 
-    __device__ __host__ inline void operator()(int x_cb, int parity)
+    __device__ __host__ inline void operator()(int x_cb, int flavor, int parity)
     {
       using namespace linalg; // for Cholesky
       const int clover_parity = arg.nParity == 2 ? parity : arg.parity;
       const int spinor_parity = arg.nParity == 2 ? parity : 0;
       constexpr int n_flavor = 2;
-      fermion in[n_flavor];
-      fermion out[n_flavor];
 
-#pragma unroll
-      for (int flavor = 0; flavor < n_flavor; flavor++) {
-        in[flavor] = arg.in(x_cb + flavor * arg.volumeCB, spinor_parity);
-        in[flavor].toRel(); // change to chiral basis here
-      }
+      int my_flavor_idx = x_cb + flavor * arg.volumeCB;
+      fermion in = arg.in(my_flavor_idx, spinor_parity);
+      in.toRel(); // change to chiral basis here
 
-#pragma unroll
-      for (int chirality = 0; chirality < 2; chirality++) {
-        // (C + i mu gamma_5 tau_3 - epsilon tau_1 )  [note: appropriate signs carried in arg.a / arg.b]
-        const complex<real> a(0.0, chirality == 0 ? arg.a : -arg.a);
+      int chirality = flavor; // relabel flavor as chirality
+      // (C + i mu gamma_5 tau_3 - epsilon tau_1 )  [note: appropriate signs carried in arg.a / arg.b]
+      const complex<real> a(0.0, chirality == 0 ? arg.a : -arg.a);
 
-        Mat A = arg.clover(x_cb, clover_parity, chirality);
+      Mat A = arg.clover(x_cb, clover_parity, chirality);
 
-        half_fermion in_chi[n_flavor] = {in[0].chiral_project(chirality), in[1].chiral_project(chirality)};
-        half_fermion out_chi[n_flavor];
+      SharedMemoryCache<half_fermion> cache(target::block_dim());
 
+      half_fermion in_chi[n_flavor]; // flavor array of chirally projected fermion
 #pragma unroll
-        for (int flavor = 0; flavor < n_flavor; flavor++) {
-          out_chi[flavor] = A * in_chi[flavor];
-          out_chi[flavor] += (flavor == 0 ? a : -a) * in_chi[flavor];
-          out_chi[flavor] += arg.b * in_chi[1 - flavor];
-        }
+      for (int i = 0; i < n_flavor; i++) in_chi[i] = in.chiral_project(i);
```

Comment:
Should we put `#pragma unroll` here just for full diligence? I know this loop should be more than simple enough for a compiler to unroll, but we may as well give it more information if we can.